spring-integration-file/src/main/java/org/springframework/integration/file/remote/synchronizer/AbstractInboundFileSynchronizer.java

```diff
@@ -436,9 +440,19 @@ else if (this.logger.isWarnEnabled()) {
 		return false;
 	}
 
-	private String generateLocalFileName(String remoteFileName) {
+	private String generateLocalFileName(String remoteDirectoryPath, String remoteFileName) {
 		if (this.localFilenameGeneratorExpression != null) {
-			return this.localFilenameGeneratorExpression.getValue(this.evaluationContext, remoteFileName, String.class);
+			if (!(this.localFilenameGeneratorExpression instanceof LiteralExpression) &&
+					this.localFilenameGeneratorExpression.getExpressionString().contains("#remoteDirectory")) {
+				EvaluationContext context = IntegrationContextUtils.getEvaluationContext(this.beanFactory);
+				context.setVariable("remoteDirectory", remoteDirectoryPath);
+				return this.localFilenameGeneratorExpression.getValue(context, remoteFileName,
```

Comment:
Looks like this call can be only once. What we definitely need here is to reconfigure a `context` variable according the `#remoteDirectory` presence. But, yeah... the solution is still cool. 👍